lib/cn/node_split.c

```diff
@@ -279,7 +279,7 @@ forward_wbt_leaf_iterator_next(struct element_source *source, void **data)
     if (iter->offset.leaf_idx == 0)
         kbr_madvise_wbt_leaf_nodes(&kblk->kb_kblk_desc, desc, MADV_WILLNEED);
 
-    *data = kblk->kb_kblk_desc.map_base + desc->wbd_first_page * PAGE_SIZE +
+    *data = (void *)kblk->kb_kblk_desc.map_base + desc->wbd_first_page * PAGE_SIZE +
```

Comment:
Probably we need a "const" iterator, like how C++ handles this situation...

Reply:
I tried.  But our iterators are mostly (maybe entirely) used by `bin_heap` which uses this type for the "get next" method:
```
typedef bool
element_source_get_next(struct element_source *source, void **data);
```
To do what you are asking requires changing `void **data` to `const void **data`.

But `data` can refer to many different things depending on what is being iterated, not all of which can be made `const`.  For example, the `data` for `c0_cursor` is a `struct kvs_cursor_element`, which is modified with each `get next`.

Reply:
By the way, I would love a solution to this. I just don't see it.  Maybe a different iterator type for keys vs all the other things we iterate over (kblocks, kvsets, bonsai_kv, etc.):
```
c0_cursor_es_make[322]         c0cur->c0cur_es = es_make(c0cur_next, 0, 0);
c0kvms_cursor_create[530]      cur->c0mc_es = es_make(c0kvms_cursor_next, c0kvms_cursor_unget, 0);
c0_kvset_iterator_init[158]    iter->c0it_handle = es_make(c0_kvset_iterator_next, c0_kvset_iterator_unget, 0);
c0_kvset_iterator_init[162]    iter->c0it_handle = es_make(c0_kvset_iterator_rnext, c0_kvset_iterator_runget, 0);
cn_cursor_es_make[1346]        cncur->cncur_es = es_make(cncur_next, 0, 0);
cn_lcur_init[157]              lcur->cnlc_es = es_make(cn_lcur_read, 0, 0);
kvset_iter_create[2597]        iter->handle.kvi_es = es_make(kvset_cursor_next, 0, 0);
kvset_iter_set_start[2684]     handle->kvi_es = es_make(kvset_cursor_next, 0, 0);
reverse_kblk_iterator_init[83] iter->es = es_make(reverse_kblk_iterator_next, NULL, NULL);
forward_wbt_leaf_iterator_init[312] iter->es = es_make(forward_wbt_leaf_iterator_next, NULL, NULL);
bonsai_iter_es_make[177]       iter->bi_es = es_make(bonsai_iter_next, 0, 0);
bonsai_ingest_iter_init[241]   iter->bii_es = es_make(bonsai_ingest_iter_next, 0, 0);
lc_cursor_es_make[1034]        cur->lcc_es = es_make(lc_cursor_next, 0, 0);
```

Reply:
I mean "in addition to", then user just choose the correct iterator.  I'm sure it's a rat hole with our implementation, far easier to do in languages that support templates :-)

Reply:
Does it make more sense to apply the cast to data instead? `wnode` can be const on the pop side of things looking at the code

Reply:
So, funny thing, doing math on a (void *) isn't actually legal C, it's another GNU hack (cuz it's typeless and hence has no size).  Of course, we do it all over the place, I am guilty as well, it's just so much more conveneient than casting to (char *), do your math, then back to (void *)...

Reply:
I am also guilty :)